topi/python/topi/x86/tensor_intrin.py

```diff
@@ -177,3 +180,95 @@ def _instr(index):
 
     with tvm.build_config(offset_factor=1, partition_const_loop=True):
         return tvm.decl_tensor_intrin(C.op, _intrin_func, binds={data:a_buffer, kernel:b_buffer})
+
+
+def dot_16x1x16_int8_int8_int32_vnni():
+    """
+    Int8 dot product by every 4 elements using AVX512VNNI Cascade Lake instructions.
+    This function takes two arrays of int8 datatype -- data[4] and
+    kernel[16][4] -- and computes a dot product of data[4] with every
+    4 elements of kernels, resulting in output[16] of int32 datatype.
+    The pseudo code is as follows.
+    .. code-block:: c
+        void dot_16x1x16_int8_int8_int32_vnni(int8 data[4], int8 kernel[16][4],
+                int32 output[16]){
+            for (int i = 0; i < 16; i++){
+                out[i] = 0;
+                for (int k = 0; k < 4; k++){
+                    out[i] += data[k] * kernel[i][k]
+                }
+            }
+        }
+
+    Physically, the kernel array sits in an AVX512 vector register and
+    the data[4] is broadcasted to another AVX512 vector register. This
+    function returns a TensorIntrin that can be used to tensorize
+    a schedule.
+
+    Returns
+    -------
+    intrin : TensorIntrin
+        The Cascade Lake int8 TensorIntrin that can be used in tensorizing schedule
+    """
+
+    int32_lanes = 16 # 16 int32 lanes in AVX512
+    num_int8_elements = 4 # 4 int8 elements in int32
+    data = tvm.placeholder((num_int8_elements,), dtype='uint8', name='data')
+    kernel = tvm.placeholder((int32_lanes, num_int8_elements), dtype='int8', name='kernel')
+    k = tvm.reduce_axis((0, num_int8_elements), name='k')
+    C = tvm.compute((int32_lanes,),
```

Comment:
In my shallow knowledge on the semantics of both VNNI and TVM generated code.
There might be better software-defined description of VNNI so that you can avoid buffer reset.

Say VNNI does something like this:
````
for (int i = 0; i < 16; ++i) {
    uint32_t sum = 0;
    for (int j = 0; j < 4; ++j)
        sum += a[i * 4 + j] * b[i * 4 + j];
    c[i] = c[i] + sum; // We do not want to set c[i] to zero, it is an accumulation
}
````

However, if we build `C.op` it will generate code like this:
````
for (int i = 0; i< 16; ++i) {
    c[i] = 0; // To emulate the semantics of VNNI, we definitely do not want this reset.
    for (int j = 0; j < 4; ++j)
        c[i] = c[i] + a[i * 4 + j] * b[i * 4 + j];
}
````

Therefore, I suggest `C` is only an intermediate result, still we need another tensor:
The whole code looks like this:
```
a = placeholder()
b = placeholder()
d = placeholder()
c = reduce(sum(a, b))
e = c + d
# binds e and d to the same buffer so that the results can be retained between invocations.
```

Reply:
@were : Thanks for pointing it out. I am not familiar with TVM part code either. Let me think about this. Background: we have recently incorporated VNNI into the FBGEMM (https://github.com/pytorch/fbgemm, VNNI part code will be publish soon). We also want to check if TVM can support VNNI and we want to do some performance comparisons between TVM, MKL-DNN, and FBGEMM.